clustering_code/edchan-kmeans.qmd

```diff
@@ -0,0 +1,946 @@
+     ```python
+     earthaccess.login(persist=True)
+     ```
+  2. Modify the following sample code to search for granules of the HLSL30 product overlapping the watershed boundary from May to October 2023 (there should be 76 granules):
+     ```python
+     results = earthaccess.search_data(
+         short_name="...",
+         cloud_hosted=True,
+         bounding_box=tuple(gdf.total_bounds),
+         temporal=("...", "..."),
+     )
+     ```
+
+```{python}
+#| deletable: false
+#| nbgrader: {cell_type: code, checksum: f44e0625c261597c1454aaefbce06406, grade: false, grade_id: search-ans, locked: false, schema_version: 3, solution: true, task: false}
+#| tags: []
+auth = earthaccess.login(strategy='netrc')
+
+if not auth.authenticated:
+    raise ValueError('Authentication failed')
+```
+
+```{python}
+granules = earthaccess.search_data(
+    short_name='HLSL30',
+    cloud_hosted=True,
+    bounding_box=tuple(study_site.total_bounds),
+    temporal=('May 2023', 'October 2023')
+) # DataGranule - Dictionary-like object
+```
+
+```{python}
+granules_df = pd.json_normalize(granules)
+granules_df.to_csv(os.path.join(cache_dir, 'HLSL30_search_results.csv'))
+```
+
+### Compile information about each granule
+
+I recommend building a GeoDataFrame, as this will allow you to plot the granules you are downloading and make sure they line up with your shapefile. You could also use a DataFrame, dictionary, or a custom object to store this information.
+
+**YOUR TASK -- Write a function that:**
+  1. For each search result:
+      1. Get the following information (HINT: look at the ['umm'] values for each search result):
+          - granule id (UR)
+          - datetime
+          - geometry (HINT: check out the shapely.geometry.Polygon class to convert points to a Polygon)
+      2. Open the granule files. I recomment opening one granule at a time, e.g. with (`earthaccess.open([result]`).
+      3. For each file (band), get the following information:
+          - file handler returned from `earthaccess.open()`
+          - tile id
+          - band number
+  2. Compile all the information you collected into a GeoDataFrame
+
+```{python}
+#| deletable: false
+#| nbgrader: {cell_type: code, checksum: cac54d8cb150546f10cde4e55590fd2b, grade: false, grade_id: cell-793e5893137bcce8, locked: false, schema_version: 3, solution: true, task: false}
+#| tags: []
+def create_polygon(granule_id, data):
+    """
+    Create a shapely Polygon a list of points embedded in a dictionary.
+    """
+    if len(data) > 1:
+        raise ValueError(f'More than one data entry in the granule {granule_id}')
+    try:
+        points = data[0]['Boundary']['Points']
+        coordinates = [(point['Longitude'], point['Latitude']) for point in points]
+        return Polygon(coordinates)
+    except:
+        print(f'Error creating polygon for granule {granule_id}')
+        print(json.dumps(data))
+        raise
+```
+
+```{python}
+granules_simplified = []
```

Comment:
For this long loop containing a nested loop, it would be helpful to have some comments on what it is doing at each step. This would help the reader understand each of the processing steps.